_posts/2017-10-11-Stairway-To-Health.md

```diff
@@ -0,0 +1,356 @@
+
+<img alt="stream.png" src="{{ '/img/stairwaytohealth/stream.png' | prepend: site.baseurl }}" class="image fit" style="margin:0px auto; max-width:800px;">
+
+In the MyThings application every sensor can have a `friendlyName1` and `friendlyName2`, we used these to specify which tower and which floor they represent. 
+The sensors send a lot more data than just the magnetic pulse counts, therefore we needed the `container` field, to be able to filter on `counter` logs only (however, we store the other messages as well, maybe for future use). 
+The `value` field is the amount of times the sensor was triggered, in other words, the actual counts. And of course a `timestamp` since we will show the data in time based graphs.
+
+The `timestamp` represents the time that the sensor has sent its message to the MyThings application, we also wanted to keep track of when our (Stairway application) has received the log, so before saving we added one extra field to store this in our database.
+
+After we defined our model/schema of our logs, it was simply adding an endpoint to our express router and our first feature was ready. 
+Well not exactly, we needed to trigger an event to refresh the data on our dashboard, but we'll get back to this later.
+
+**The Dashboard**<br>
+Since we created an Angular(4) application, we took advantage of the great features of angular-cli which makes it really easy to get a new project up and running and generate new components, services, tests and much more. 
+We started by adding all the components needed for the application and adding the Proximus styling to the project. 
+After that we imported the Highcharts library from NPM to first make the charts on the homepage and later making the charts for the detailed views. 
+All the charts where first made with mock data so that we perfectly could say from the backend what data we needed and in which format. 
+From now on we knew how our JSON for the charts has to be made and we could implement the api endpoints for the dashboard and the details page. 
+Finally after adding all the charts we started on adding the different languages to the application. Here we got our biggest 'lesson learned', it is much faster to start with I18N then to end with it, this is because you have to find all the normal text in the HTML files and copy them to the JSON-files. 
+ALso we had to fast create a translation list that the business could translate for us. 
+
+
+**Mongo Aggregates**<br>
+So for displaying the daily, weekly and total counts below the buildings, we had to get this data from the database, keeping in mind that we would have to iterate over millions of sensor logs (at the time of writing this blog post 1.4 million over 4 months). We had to make sure it was performant. This is where the mongo aggregates come in handy. In stead of (say) looping over the results and adding them up, we let mongo take care of this with the '$sum' operator which in code looks like the following:
```

Comment:
**As** for displaying ...
this blog post**,** 1.4 million over 4 months (comma)
Mongo instead of mongo?
**Instead** of
Backticks for `$sum`